### ISSUE TITLE
Broken function when looping over structures with pointers-to-members

### ISSUE THREAD

**Issue by arichardson** [CTSRD-CHERI/llvm-project#555]
https://cheri-compiler-explorer.cl.cam.ac.uk/z/vjd7YP
```

Offset not zero at the point of scalar access
  %a.sroa.11.0.a.sroa.11.0.a.sroa.11.0.a.sroa.11.0.a.sroa.11.48..fca.1.load = load i64, i64 addrspace(200)* %a.sroa.11.0..fca.1.gep4.sroa_cast55, align 16, !dbg !67, !tbaa !69
!69 = !{!70, !65, i64 16}
16
in function _ZN1c1dEv
fatal error: error in backend: Broken function found, compilation aborted!
```

```c++
// RUN: %clang_cc1 -triple riscv64-unknown-freebsd13 --mrelax-relocations -fhalf-no-semantic-interposition -target-feature +m -target-feature +a -target-feature +f -target-feature +d -target-feature +c -target-feature +xcheri -target-feature -relax -target-feature +cap-mode -target-feature -save-restore -target-abi l64pc128d -msmall-data-limit 0 --compress-debug-sections=zlib -fcoverage-compilation-dir=/local/scratch/alr48/cheri/build/plasma-workspace-riscv64-purecap-build -O2 -std=gnu++17 -fvisibility-inlines-hidden -fno-operator-names -fno-signed-char -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o - -Wimplicit-int  -Wfatal-errors -S %s
namespace a {
class b;
}
class c {
  void m_fn1(int (a::b::*)() const);
  void d();
};
using namespace a;
void c::d() {
  struct {
    void (b::*e)();
    int (b::*f)() const;
  } g[]{{}};
  for (auto a : g) {
    m_fn1(a.f);
    ({ a; });
  }
}
```

**Comment by arichardson** [comment]
And another similar crash but this time without member pointers: https://cheri-compiler-explorer.cl.cam.ac.uk/z/v19ovx

```
// RUN: %clang_cc1 -triple riscv64-unknown-freebsd13 --mrelax-relocations -fhalf-no-semantic-interposition -target-feature +m -target-feature +a -target-feature +f -target-feature +d -target-feature +c -target-feature +xcheri -target-feature -relax -target-feature +cap-mode -target-feature -save-restore -target-abi l64pc128d -msmall-data-limit 0 --compress-debug-sections=zlib -fcoverage-compilation-dir=/local/scratch/alr48/cheri/build/libkscreen-riscv64-purecap-build -O2 -std=gnu++17 -fvisibility-inlines-hidden -fno-operator-names -fno-signed-char -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o - -Wimplicit-int  -Wfatal-errors -S %s
int a;
class g {
public:
  int c();
  void e();
};
class h {
public:
  bool e() {
    int b = atomic.c();
    if (b == 0)
      return false;
    if (b != -1)
      atomic.e();
    return true;
  }
  g atomic;
};
struct i {
  h e;
};
class j {
public:
  j(char *);
};
class k {
public:
  k(j);
  k(const k &) noexcept;
  void operator+=(k &);
  i d;
};
k::k(const k &) noexcept { d.e.e(); }
k operator+(k m, k p2) {
  k l(m);
  l += p2;
  return l;
}
template <class> class o;
class H;
class p {
public:
  class J {
  public:
    o<H> operator*();
    bool operator!=(J);
    void operator++();
  };
  J begin();
  J end();
};
template <class n> class o {
public:
  n *operator->() const;
};
class r {
public:
  p outputs();
};
class t {
public:
  t(int *);
  void operator<<(k);
};
class H {
public:
  bool q();
};
class M {
  void v() const;
  o<r> s;
};
t c(&a);
k d(0), e(0);
void M::v() const {
  for (auto f : s->outputs())
    c << (f->q() ? d + j("") : e + j(""));
}
```

**Comment by jrtc27** [comment]
These look like the TBAA SROA issues fixed upstream last year; likely d3e2b4c0a21f8b3861d8b64a4c2ed051694094d1 specifically (which I had to cherry-pick to Morello LLVM 12.x to fix building Qt, presumably also the origin of this reduced C++ source). No longer reproducible.
